### ISSUE TITLE
Generic constructor inference becomes extremely slow when the expected type is a recursive protocol

### ISSUE THREAD

**Issue by OutSquareCapital** [astral-sh/ty#4269]
### Summary

Tested with last version from the root of [my repo](https://github.com/OutSquareCapital/pyochain)

No crash from basedpyright, pyrefly, nor mypy when checking all files.

with `uvx ty check tests/benchmarks/`, it take like 1 min to check 14 files. It hang on the last one.
with `uvx ty check tests/test_abc`, it issues a `WARN, a fatal error occured`
same thing with `uvx ty check tests/test_extern/`
same thing with `uvx ty check tests/test_cpython/` (adapted from official python test suite)
same thing with `uvx ty check tests/test_last_count.py`

Haven't checked more. `test_last_count` is like 26 lines, it could be a good start to investiguate what happens.

No need to mention that trying to run the language server is not feasible.

It even crashes on this:
```python
from pyochain import Dict, Vec

x = (
    Vec([1, 2, 3, 4, 5])
    .iter()
    .enumerate()
    .map_star(lambda i, x: (i, x * 2))
    .collect(Dict)
)
```

See [Dict](https://github.com/OutSquareCapital/pyochain/blob/master/pyochain/core/_dict.pyi) and [Vec](https://github.com/OutSquareCapital/pyochain/blob/master/pyochain/core/_vec.pyi) stubs.

basedpyright correctly infer the return type as `Dict[int, int]` (pyrefly don't crash but issues false warnings).

with just `Vec([1,2,3,4,5]).iter().collect(list)`, pyrefly is fine but ty still crashes

I wish that lambdas support would be more prioritized :) I can't use anything besides basedpyright so far unfortunately.

### Version

ty 0.0.71 (683e22331 2026-08-12)

**Comment by carljm** [comment]
Thanks for the report!

The crash is fixed by https://github.com/astral-sh/ruff/pull/27737.

The performance problem is separate; we'll use this issue to track it.

**Comment by carljm** [comment]
It looks like the performance issues here have a couple related causes, but will be most clearly addressed in a series of separate PRs -- https://github.com/astral-sh/ruff/pull/27772 is the first pre-requisite change.

**Comment by carljm** [comment]
@OutSquareCapital This project is proving to be quite a useful stress test! Several related fixes have landed, more are in the works.

How would you feel about us adding pyochain to https://github.com/hauntsaninja/mypy_primer so that we type check it as part of our CI?

**Comment by OutSquareCapital** [comment]
It would be an honor to be added to ty mypy-primer list 😄.

Coincidentally, I worked yesterday on a refactor of my dedicated [typing check file](https://github.com/OutSquareCapital/pyochain/blob/dev/tests/test_typing.py), who now uses `typing::assert_type` whenever possible.

I would be very keen to expand/adjust it, I think there's a few interesting things to add that would be good stress tests for ty.

Only looking at `PyoIterator` for example, I could check methods like [`PyoIterator::map_star`](https://github.com/OutSquareCapital/pyochain/blob/3605b97e57b8e66c0a9e10d9b7cfcb1d4246a787/pyochain/abc/_iterator.pyi#L2555), [`PyoIterator::fold_star`](https://github.com/OutSquareCapital/pyochain/blob/3605b97e57b8e66c0a9e10d9b7cfcb1d4246a787/pyochain/abc/_iterator.pyi#L1733), add more edge cases for [`PyoIterator::flatten`](https://github.com/OutSquareCapital/pyochain/blob/3605b97e57b8e66c0a9e10d9b7cfcb1d4246a787/pyochain/abc/_iterator.pyi#L1586) and [`PyoIterator::chain`](https://github.com/OutSquareCapital/pyochain/blob/3605b97e57b8e66c0a9e10d9b7cfcb1d4246a787/pyochain/abc/_iterator.pyi#L777), or even some that may seem simple at first, like [`PyoIterator::collect`](https://github.com/OutSquareCapital/pyochain/blob/3605b97e57b8e66c0a9e10d9b7cfcb1d4246a787/pyochain/abc/_iterator.pyi#L816), but that ty couldn't handle last time I checked (see my comment on the pyrefly repo [here](https://github.com/facebook/pyrefly/issues/2450#issuecomment-4528334201)).

The only friction I see is that there's no way, to my knowledge, to assert expected type checking errors, especially regardless of the type checker used. 

What I'm doing currently is to use basedpyright as a baseline, add ignore comments when it seem like it *should* be a type checking error , and then rechecking with pyrefly to add their ignore comments only at places where basedpyright complains. 

This often happens in the adapted CPython test suite for example, where there's lot of errors tests for erroneous inputs, or when the closest equivalent in typeshed also has type ignore comments.

I was planning to do the same for ty, but that's when I encountered the issues reported here. 

Unfortunately it's very time consuming and hence only partially done for pyrefly, not really readable (see [here](https://github.com/OutSquareCapital/pyochain/blob/3605b97e57b8e66c0a9e10d9b7cfcb1d4246a787/tests/test_typing.py#L273)), and there's no way to clearly indicate expected errors besides comments (e.g [here](https://github.com/OutSquareCapital/pyochain/blob/3605b97e57b8e66c0a9e10d9b7cfcb1d4246a787/tests/test_typing.py#L270), basedpyright correctly complains but pyrefly don't).

I tought about adding a script to automatically add corresponding ignore comments for both ty and pyrefly, but now I have to maintain the script, a table of correspondances, etc...

Finally, the [upcasting checks](https://github.com/OutSquareCapital/pyochain/blob/3605b97e57b8e66c0a9e10d9b7cfcb1d4246a787/tests/test_typing.py#L76), i.e => "does `x: Foo[Subclass]` can be accepted as `x: Foo[ParentClass]`?" can only be implicit, since `typing::assert_type` expect the most precise types only.

If you have a few ideas to improve this so that ty can benefit from it once added to the mypy-primer list, I'm all ears!

The best time to add it would be on my next release since it will land with the refactored typing check file, I'll notify once that's done.

In any case I'm looking forwards to the upcoming fixes :)

**Comment by carljm** [comment]
Writing a cross-type-checker "typing test suite" is indeed somewhat painful, I think. Even `assert_type` is often over-constraining, since in many cases there can be multiple valid solutions, and type checkers may have good reasons for picking different ones. I think it's usually better to test expected assignability results than to use `assert_type`.

As far as suppressions, I think your choices are basically to either use blanket `# type: ignore` comments (without a rule code), or to add a suppression comment per type checker (as you are currently doing). Another issue that can crop up is type checkers emitting some errors in different locations.

Regarding `PyoIterator::collect`, we do better there now than we did back in May (we correctly infer all cases in your example that provide an explicit argument); the remaining issue looks like #592 -- this is something that is still only supported by pyright and basedpyright.

I'll add pyochain to mypy-primer after we are able to type check it without slowdown :) But before closing this issue.

**Comment by jelle-openai** [comment]
I might at some point write a generalization of the python/typing conformance suite runner that can be used to write typing tests for your own projects; that would hopefully work well for you here.

**Comment by OutSquareCapital** [comment]
> As far as suppressions, I think your choices are basically to either use blanket `# type: ignore` comments (without a rule code), or to add a suppression comment per type checker (as you are currently doing). Another issue that can crop up is type checkers emitting some errors in different locations.

Mhm I see. I'll keep using the "suppression comments per type checker" then, so fixed false positives/new true flagged errors are always notified.

> Regarding `PyoIterator::collect`, we do better there now than we did back in May (we correctly infer all cases in your example that provide an explicit argument); the remaining issue looks like [#592](https://github.com/astral-sh/ty/issues/592) -- this is something that is still only supported by pyright and basedpyright.

Nice!
> I'll add pyochain to mypy-primer after we are able to type check it without slowdown :) But before closing this issue.

Okay, make sense



> I might at some point write a generalization of the python/typing conformance suite runner that can be used to write typing tests for your own projects; that would hopefully work well for you here.

Oh yeah that would be very interesting. Feel free to notify me when/if you start working on that

**Comment by OutSquareCapital** [comment]
Hey @carljm , an underlying issue that may lead to a lot of divergence between my "baseline" basedpyright and ty is that the former accept `PyoIterator` and `Iter` as covariant whilst not ty nor pyrefly do. I'm not sure who's right tho.
I tried smaller reproducible examples but they were working as expected (i.e all 3 checkers converging).

Examples are below. Do you want me to raise another issue? 

---

## First example

```python
from __future__ import annotations

from typing import TYPE_CHECKING, assert_type

from pyochain import Iter

if TYPE_CHECKING:
    from collections.abc import Iterator

    from pyochain.abc import PyoIterator


class Animal: ...


class Dog(Animal): ...


a = assert_type(Iter(Dog()), Iter[Dog])
_b: Iter[Animal] = a
_c: PyoIterator[Animal] = a
_d: Iterator[Animal] = a
```
### ty output
```python
PS C:\Users\stett\Documents\python\pyochain> uvx ty check t.py
error[invalid-assignment]: Object of type `Iter[Dog]` is not assignable to `Iter[Animal]`
  --> t.py:20:20
   |
20 | _b: Iter[Animal] = a
   |     ------------   ^ Incompatible value of type `Iter[Dog]`
   |     |
   |     Declared type
info: `Iter` is invariant in its type parameter
info: For more information, see https://docs.astral.sh/ty/reference/typing-faq/#invariant-generics

error[invalid-assignment]: Object of type `Iter[Dog]` is not assignable to `PyoIterator[Animal]`
  --> t.py:21:27
   |
21 | _c: PyoIterator[Animal] = a
   |     -------------------   ^ Incompatible value of type `Iter[Dog]`
   |     |
   |     Declared type
info: type `Iter[Dog]` is not assignable to protocol `PyoIterator[Animal]`
info: └── protocol member `batched` is incompatible

Found 2 diagnostics
```
### Pyrefly output
```shell
PS C:\Users\stett\Documents\python\pyochain> uvx pyrefly check t.py  
ERROR `Iter[Dog]` is not assignable to `Iter[Animal]` [bad-assignment]
  --> t.py:20:20
   |
20 | _b: Iter[Animal] = a
   |     ------------   ^
   |     |
   |     declared type
   |
ERROR `Iter[Dog]` is not assignable to `PyoIterator[Animal]` [bad-assignment]
  --> t.py:21:27
   |
21 | _c: PyoIterator[Animal] = a
   |     -------------------   ^
   |     |
   |     declared type
   |
 INFO 2 errors
```
### Basedpyright output
```shell
PS C:\Users\stett\Documents\python\pyochain> uv run basedpyright t.py
0 errors, 0 warnings, 0 notes
PS C:\Users\stett\Documents\python\pyochain> 
```
## Second example

Replacing the iterators by sequences (may be easier to pinpoint where it comes from, since the number of methods is much smaller)
```python
# Swap from pyochain import Iter to from pyochain import Seq
a = assert_type(Seq(Dog()), Seq[Dog])
_b: Seq[Animal] = a
_c: PyoSequence[Animal] = a
_d: Sequence[Animal] = a
```
### ty output
```shell
PS C:\Users\stett\Documents\python\pyochain> uvx ty check t.py       
error[invalid-assignment]: Object of type `Seq[Dog]` is not assignable to `Seq[Animal]`
  --> t.py:20:19
   |
20 | _b: Seq[Animal] = a
   |     -----------   ^ Incompatible value of type `Seq[Dog]`
   |     |
   |     Declared type
info: `Seq` is invariant in its type parameter
info: For more information, see https://docs.astral.sh/ty/reference/typing-faq/#invariant-generics

error[invalid-assignment]: Object of type `Seq[Dog]` is not assignable to `PyoSequence[Animal]`
  --> t.py:21:27
   |
21 | _c: PyoSequence[Animal] = a
   |     -------------------   ^ Incompatible value of type `Seq[Dog]`
   |     |
   |     Declared type

Found 2 diagnostics
```
### pyrefly output
```shell
PS C:\Users\stett\Documents\python\pyochain> uvx pyrefly check t.py
ERROR `Seq[Dog]` is not assignable to `Seq[Animal]` [bad-assignment]
  --> t.py:20:19
   |
20 | _b: Seq[Animal] = a
   |     -----------   ^
   |     |
   |     declared type
   |
ERROR `Seq[Dog]` is not assignable to `PyoSequence[Animal]` [bad-assignment]
  --> t.py:21:27
   |
21 | _c: PyoSequence[Animal] = a
   |     -------------------   ^
   |     |
   |     declared type
   |
 INFO 2 errors
```
### basedpyright output
```shell
PS C:\Users\stett\Documents\python\pyochain> uv run basedpyright t.py
0 errors, 0 warnings, 0 notes
PS C:\Users\stett\Documents\python\pyochain> 
```

**Comment by carljm** [comment]
Yes, separate issue will be easier to manage; I created https://github.com/astral-sh/ty/issues/4407

**Comment by carljm** [comment]
ty 0.0.78 now checks pyochain in a reasonable amount of time; roughly on par with pyright. I think we should be able to still do quite a bit better here, but as a blocking hang, I think we can consider this issue resolved.

**Comment by carljm** [comment]
It looks like almost all of our time is spent checking `tests/test_typing/check_typing.py`, so that suggests there is still a pathological performance issue (just on a lesser scale) that we should be able to fix to gain another dramatic speedup here.

**Comment by carljm** [comment]
Going to reopen this issue to track the remaining perf fix, as well as adding pyochain to mypy-primer (and ideally our ty project benchmarks, too).

**Comment by OutSquareCapital** [comment]
Yep, I can confirm it does finish on the master branch!

But on the [dev branch](https://github.com/OutSquareCapital/pyochain/tree/dev), it struggle on the 112th file, and never goes past the 113th (I waited +30 sec) if I run `uvx ty check .` at the root.

The most likely candidates to this issue are a few new typing tests in [`tests/test_typing/check_iter.py`](https://github.com/OutSquareCapital/pyochain/blame/5ae86dd99cb704e2568ceee3b5940de454002619/tests/test_typing/check_iter.py#L101), and the refactor of some overload signatures in [`PyoIterator::for_each_star`](https://github.com/OutSquareCapital/pyochain/blob/5ae86dd99cb704e2568ceee3b5940de454002619/pyochain/abc/_iterator.pyi#L1800) and [`PyoIterator::map_windows_star`](https://github.com/OutSquareCapital/pyochain/blob/5ae86dd99cb704e2568ceee3b5940de454002619/pyochain/abc/_iterator.pyi#L2765).

There's some noise in the change of the `PyoIterator` file because I changed all the `self: PyoIterator[Foo]` bounds to `self: Iterator[Foo]`, but I doubt it's related.

**Comment by carljm** [comment]
Looking into the perf issues on the dev branch.

**Comment by OutSquareCapital** [comment]
@carljm  I just retested now on dev branch, and it's now as fast as I expect ty to be! 

Tested with `uvx ty check .`, or with focused tests like `uvx ty check tests/test_typing/`.

There's a ton of issues non-flagged by basedpyright but that's another story, hence to me this can be considered as resolved :).

EDIT: Altough ty is now much faster than basedpyright, it's still noticeably slower than pyrefly, even when checking "only" the `test_typing` folder. So there's still probably some room for improvement, but the "usable" milestone is attained.

**Comment by carljm** [comment]
Thanks @OutSquareCapital! I think we should be as fast as pyrefly here (and last I tested we were on main branch). It's great to hear that some other changes in ty have made dev branch usable, but I still have some unpublished changes locally that I think should significantly improve perf on dev branch too, so I'll keep this open until those are pushed, reviewed, and landed also.
